File: solutions/graphs/bfs/employeeImportance.js
const buildEmployeeMap = (employees) => {
    const idToEmployeeMap = {};

    for (let i = 0; i < employees.length; i++) {
        const curEmployee = employees[i];
        idToEmployeeMap[curEmployee.id] = curEmployee;
    }

    return idToEmployeeMap;
};

const getTotalImportance = (rootId, idToEmployeeMap) => {
    let totalImportance = 0;

    const queue = new Queue();
    queue.enqueue(rootId);

    while (queue.size() > 0) {
        // Remove node
        const employeeId = queue.dequeue();

        // Process node
        const employee = idToEmployeeMap[employeeId];
        const { id, importance, subordinates } = employee;

        totalImportance += importance;

        // Add neighbors
        subordinates.forEach((subordinate) => {
            // This is a DAG (directed acyclic graph) meaning we do NOT need any visited checks.
            // What would a cycle mean in this context?
            // Imagine that Bob is your manager, and Abby is Bob's manager, but you are also
            // Abby's manager. Clearly, that does not make sense (because you would be in
            // charge of your own grandmanager). So, we see that in this employee hierarchy
            // graph specified for this problem, cycles are not possible. Thus, we do NOT
            // need any visited checks.
            queue.enqueue(subordinate);
        });
    }

    return totalImportance;
};

const GetImportance = function (employees, id) {
    const idToEmployeeMap = buildEmployeeMap(employees);

    return getTotalImportance(id, idToEmployeeMap);
};

File: solutions/python/graphs/bfs/employee_importance.py
class Solution:
    def build_employee_map(self, employees):
        id_to_employee_map = {}
        
        for employee in employees:
            id_to_employee_map[employee.id] = employee
        
        return id_to_employee_map
    
    def get_total_importance(self, root_id, id_to_employee_map):
        total_importance = 0
        
        queue = deque()
        queue.append(root_id)
        
        while queue:
            # Remove node
            employee_id = queue.popleft()
            
            # Process node
            employee = id_to_employee_map[employee_id]
            
            total_importance += employee.importance
            
            # Add neighbors
            for subordinate in employee.subordinates:
                queue.append(subordinate)
                
        return total_importance
    
    def getImportance(self, employees: List['Employee'], id: int) -> int:
        id_to_employee_map = self.build_employee_map(employees)
        
        return self.get_total_importance(id, id_to_employee_map)
        